### PR TITLE
Implement the Themes REST Service

### PR DESCRIPTION
**Pull request by BaguetteEater**
Themes can now get fetch via the route "/theme/:id" where :id is the id of the wanted theme.
Theme is a es6 class which increase code readability and manipulation. All its parameters can be getted and setted for now.

The Colors class is unsed until further explanation about its structure and attributes.

Connexions with the database is made possible with NodePosgres and its documentation can be found [here](https://node-postgres.com/).

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+);
+
+app.use('/theme', themeRouter);
+
+app.use((req, res, next) => {
+    res.setHeader('Content-Type', 'text/plain');
+    res.status(404).send('Page introuvable !');
+});
+
+app.listen(8079, () => {
```
Comment (aarrtthhuurr) [inline index.js]: Maybe define the Port in environement variables (or a config file)

File: routes/theme.js [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+        }).catch((err) => {
+            res.status(500).send(errfactory.errorMessageToJson(err, errMessages.db.theme));
+        });
+});
+
+router.post('/:id', (req, res) => {
+    res.send("lol");
+});
+
+module.exports = router;
```
Comment (aarrtthhuurr) [inline routes/theme.js]: Add a blank line at the end of your files

File: routes/error_messages_factory.js [hunk 3]
```diff
@@ -0,0 +1,3 @@
+module.exports.errorMessageToJson = (err, text) => {
```
Comment (aarrtthhuurr) [inline routes/error_messages_factory.js]: - Name files in camelCase
- Chose between `errorMessages` or `errorMessage` (I would prefer singular)

File: routes/theme.js [hunk 4]
```diff
@@ -0,0 +1,25 @@
+const express = require('express');
+const db = require('../db/theme');
+const router = express.Router();
+
+const { Theme } = require('../model/theme');
+const errMessages = require('../errorsMessages');
+const errfactory = require('./error_messages_factory');
+
+router.get('/:id', (req, res) => {
+//TODO à alleger : c'est possible mais j'ai la flemme d'opti
```
Comment (aarrtthhuurr) [inline routes/theme.js]: Avoid comments like this

File: model/colors.js [hunk 5]
```diff
@@ -0,0 +1,43 @@
+class Colors{
```
Comment (aarrtthhuurr) [inline model/colors.js]: `class Colors {`

File: model/theme.js [hunk 6]
```diff
@@ -0,0 +1,23 @@
+class Theme {
+    constructor(row) {
+
```
Comment (aarrtthhuurr) [inline model/theme.js]: Useless blank line

File: model/theme.js [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+        this.name = row.themename;
+        this.beginDate = row.begindate;
+        this.endDate = row.enddate;
+        this.colors = JSON.parse(row.colors);
+
```
Comment (aarrtthhuurr) [inline model/theme.js]: Useless blank line

File: database.sql [hunk 8]
```diff
@@ -0,0 +1,27 @@
+create table if not exists theme (
```
Comment (aarrtthhuurr) [inline database.sql]: SQL is not matching the classes

**Review by aarrtthhuurr — CHANGES_REQUESTED** [review]

File: .gitignore [hunk 9]
```diff
@@ -0,0 +1,4 @@
+.idea
+node_modules/
+db_config.json
+config/
```
Comment (aarrtthhuurr) [inline .gitignore:4]: Can you add a not-ignored file named `config.example` describing variables?

File: package.json [hunk 10]
```diff
@@ -0,0 +1,42 @@
+{
+  "name": "payback",
+  "version": "1.0.0",
+  "description": "The RESTful server dedicated to serve the PayUTC mobile app",
+  "main": "index.js",
+  "scripts": {
+    "test": "mocha --timeout 10000",
+    "lint": "eslint .. La date de parution",
```
Comment (aarrtthhuurr) [inline package.json]: Is it a real command?

File: sql/v2.sql [hunk 11]
```diff
@@ -0,0 +1,9 @@
+alter table theme (
```
Comment (aarrtthhuurr) [inline sql/v2.sql]: No v2 required yet. Juste update the v1 file.

File: sql/v1.sql [hunk 12]
```diff
@@ -0,0 +1,26 @@
+create table if not exists theme (
```
Comment (aarrtthhuurr) [inline sql/v1.sql]: Rename this file explaining what it is doing: `v1_create_theme_table` for instance

File: db/theme.js [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+let getAllThemes = () => {
+    return pool.query('select * from theme');
+};
+
+let getTheme = (id) => {
+    return pool.query('select * from theme where $1 = id', [id]);
+};
+
+let postTheme = (theme) => {
+    let text = "INSERT INTO theme(themename, begindate, enddate, colors) VALUES($1, $2, $3, $4)";
```
Comment (aarrtthhuurr) [inline db/theme.js]: Chose between capitalized SQL or not (I'd recommand to capitalize every sentences)

File: db/theme.js [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+
+const pool = new Pool({
+    user: config.user,
+    host: config.host,
+    database: config.database,
+    password: config.password,
+    port: config.port,
+});
+
+let getAllThemes = () => {
```
Comment (aarrtthhuurr) [inline db/theme.js]: There is a lot a `let` which could be `const` in your project

File: index.js [hunk 15]
```diff
@@ -0,0 +1,27 @@
...
+    bodyParser.urlencoded({
+        extended: false,
+    })
+);
+
+app.use('/theme', themeRouter);
+
+app.use((req, res, next) => {
+    res.setHeader('Content-Type', 'text/plain');
+    res.status(404).send('Page introuvable !');
```
Comment (aarrtthhuurr) [inline index.js]: English everywhere

**Review by aarrtthhuurr — CHANGES_REQUESTED** [review]

File: package.json (hunk 10 above)
Comment (BaguetteEater) [reply package.json]: npm init does not like when you specify that you want to implement eslint without installing it, it is changed

**Review by aarrtthhuurr — APPROVED** [review]

**Pull request merged**